internal/venice-client-common/src/main/java/com/linkedin/venice/compression/ReusableGzipOutputStream.java

```diff
@@ -0,0 +1,136 @@
+package com.linkedin.venice.compression;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.zip.CRC32;
+import java.util.zip.Deflater;
+import java.util.zip.DeflaterOutputStream;
+
+
+/**
+ * This class implements a stream filter for writing compressed data in the GZIP file format. It's an adoption of
+ * {@link java.util.zip.GZIPOutputStream} but with a notable difference regarding re-usability:
+ * <ul>
+ * <li>Expose {@link #reset()} to reset CRC32 and the deflater</li>
+ * <li>Don't write the GZIP header upon construction but expose {@link #writeHeader()}</li>
+ * </ul>
+ */
+public class ReusableGzipOutputStream extends DeflaterOutputStream {
+  /**
+   * CRC-32 of uncompressed data.
+   */
+  private final CRC32 crc = new CRC32();
+
+  /*
+   * GZIP header magic number.
+   */
+  private final static int GZIP_MAGIC = 0x8b1f;
+
+  /*
+   * Trailer size in bytes.
+   *
+   */
+  private final static int TRAILER_SIZE = 8;
+
+  /**
+   * Creates a new output stream with the specified buffer size and flush mode.
+   *
+   * @param out the output stream
+   * @exception IllegalArgumentException if {@code size <= 0}
+   *
+   * @since 1.7
+   */
+  public ReusableGzipOutputStream(OutputStream out) {
```

Comment:
Just wondering whether is it possible to extend GZIPOutputStream, reset the written header in this constructor and add reset() method to it as all other method looks the same.

Reply:
Could be done, but still needs couple more methods to the copied over, like ` private void writeHeader() ` is private in GZIPOutputStream which cannot be used.

Reply:
Got it, I think it's better this way rather than copying half of it and letting the other half remain in the library: might create issues if we upgrade the library. Resolving it, thanks.